web/src/channel_folders_ui.ts

```diff
@@ -258,6 +282,126 @@ function get_edit_modal_folder_id_if_open(): number | undefined {
     return Number($edit_folder_modal.find(".stream-list-container").attr("data-folder-id"));
 }
 
+function get_channel_folder_candidates(folder_id: number): dropdown_widget.Option[] {
+    return stream_data
+        .get_unsorted_subs()
+        .flatMap((stream) =>
+            stream.folder_id !== folder_id
+                ? [
+                      {
+                          name: stream.name,
+                          unique_id: stream.stream_id,
+                          stream,
+                      },
+                  ]
+                : [],
+        )
+        .toSorted((a, b) => {
+            if (a.name.toLowerCase() < b.name.toLowerCase()) {
+                return -1;
+            }
+            if (a.name.toLowerCase() > b.name.toLowerCase()) {
+                return 1;
+            }
+            return 0;
+        });
+}
+
+function get_channel_folder_candidates_for_dropdown(): dropdown_widget.Option[] {
+    const folder_id = get_edit_modal_folder_id_if_open();
+    assert(folder_id);
+    return get_channel_folder_candidates(folder_id);
+}
+
+function change_state_of_subscribe_button(
+    event: JQuery.ClickEvent,
+    dropdown: tippy.Instance,
+): void {
+    dropdown.hide();
+    event.preventDefault();
+    event.stopPropagation();
+    assert(add_channel_folder_widget !== undefined);
+    add_channel_folder_widget.render();
+    $("#edit_channel_folder .add-channel-button").prop("disabled", false);
+}
+
+function reset_add_channel_widget(): void {
+    $("#edit_channel_folder .add-channel-button").prop("disabled", true);
+
+    $("#add_channel_folder_widget .dropdown_widget_value").text(
+        $t({defaultMessage: "Select a channel"}),
+    );
+
+    if (add_channel_folder_widget) {
+        add_channel_folder_widget.current_value = undefined;
+    }
+}
+
+function render_add_channel_folder_widget(): void {
+    const opts: DropdownWidgetOptions = {
+        widget_name: "add_channel_folder",
+        get_options: get_channel_folder_candidates_for_dropdown,
+        item_click_callback: change_state_of_subscribe_button,
+        $events_container: $("#edit_channel_folder"),
+        unique_id_type: "number",
+    };
+    add_channel_folder_widget = new dropdown_widget.DropdownWidget(opts);
+    add_channel_folder_widget.setup();
+
+    const $add_channel_button = $("#edit_channel_folder .add-channel-button");
+    $("#add_channel_folder_widget .dropdown_widget_value").text(
+        $t({defaultMessage: "Select a channel"}),
+    );
+
+    $add_channel_button.prop("disabled", true);
+
+    $add_channel_button.on("click", (e) => {
+        e.preventDefault();
+        assert(add_channel_folder_widget);
```

Comment:
These asserts should be complete like `assert(add_channel_folder_widget !== undefined)`